### PR TITLE
fix(upgrade): fix empty transclusion content with AngularJS@>=1.5.8

### LINKED ISSUE #22175: Upgraded components with ng-transclude throw an error if there is no content
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
When upgrading Angular JS component that uses transclude option an error is thrown at runtime as it attempts to destroy transcluded scope. This is only happening when no content is passed.
image

Aside of that, it appears that `$transclude` local is not available either and thus cannot be injected ? 

Related issue: #13271 and possible PR #22167 

## Expected behavior

No error should not be thrown

## Minimal reproduction of the problem with instructions
https://stackblitz.com/edit/ngupgradestatic-playground-hh5hru?file=ng1%2Fng1-demo.component.ts

The error will be visible in console. 

## Environment

Angular version: 5.2.4
AngularJS version: 1.6.9

Browser: any

### PR DESCRIPTION
**Pull request by gkalpak**
# PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

## PR Type
```
[x] Bugfix
[x] Build related changes
```

## What is the current behavior?
The function provided by `ngUpgrade` as `parentBoundTranscludeFn` when upgrading a component with transclusion, will break in AngularJS v1.5.8+ if no transclusion content is provided. The reason is that AngularJS will try to destroy the transclusion scope (which would not be needed any more). But since the transcluded content comes from Angular, not AngularJS, there is no transclusion scope to destroy.

Issue Number: #22175

## What is the new behavior?
Transclusion works even if no content is provided with AngularJS v1.5.8+. `ngUpgrade` provides a dummy scope object with a no-op `$destroy()` method to avoid throwing an error.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Fixes #22175.
This commit also upgrades AngularJS to the latest stable version.

### CODE REVIEW

**Review by petebacondarwin — APPROVED** [review]
WOT NO TEST?

**Comment by gkalpak** [comment]
The tests we had were sufficient (as long as we use an AngularJS version >=1.5.8). Upgrading AngularJS was my test 😛
<sub>(But now some other tests fail. Fixing...)</sub>

**Comment by gkalpak** [comment]
I managed to get karma to run tests for multiple AngularJS versions (currently 1.5.x and 1.6.x) and then I realized I have no idea what to do with bazel 😞

@petebacondarwin, feel free to PTAL anyway :grin:
(The first 4 commits are new. You have already reviewed the 5th one.)

File: packages/upgrade/src/common/upgrade_helper.ts [hunk 1]
```diff
@@ -123,8 +123,14 @@ export class UpgradeHelper {
     const transclude = this.directive.transclude;
     const contentChildNodes = this.extractChildNodes();
     let $template = contentChildNodes;
-    let attachChildrenFn: angular.ILinkFn|undefined = (scope, cloneAttach) =>
-        cloneAttach !($template, scope);
+    let attachChildrenFn: angular.ILinkFn = (scope, cloneAttachFn) => {
```
Comment (alfaproject) [inline packages/upgrade/src/common/upgrade_helper.ts]: It still can be `undefined`, right?

Comment (gkalpak) [reply packages/upgrade/src/common/upgrade_helper.ts]: I don't see how 😕 (TypeScript agrees :smiley:) 
It is assigned a (defined) value and is never re-assigned. It should actually be `const` :grin:

**Review by petebacondarwin — COMMENTED** [review]
In `refactor(upgrade): use correct paths for imports`

Should the word `alias` really be `symlink`?

File: packages/upgrade/src/common/angular1.ts [hunk 2]
```diff
@@ -289,4 +291,4 @@ export const resumeBootstrap = () => angular.resumeBootstrap();
 
 export const getTestability = (e: Element) => angular.getTestability(e);
 
-export const version = angular.version;
+export let version = angular.version;
```
Comment (petebacondarwin) [inline packages/upgrade/src/common/angular1.ts]: By setting this to let, will it automatically update when the AngularJS global is updated?

File: packages/upgrade/test/BUILD.bazel [hunk 3]
```diff
@@ -23,8 +23,12 @@ ts_web_test(
     name = "test_web",
     bootstrap = [
         "//:web_test_bootstrap_scripts",
-        "//:angularjs",
+        # "//:angularjs",
     ],
+    # Disable since tests need to request different AngularJS versions at
+    # runtime, which is not yet supported.
+    # (Related issue: https://github.com/bazelbuild/rules_typescript/issues/131)
+    tags = ["manual"],
```
Comment (petebacondarwin) [inline packages/upgrade/test/BUILD.bazel]: Do we need some sort of TODO in the code here, or an issue in the Angular repo, to track this?

File: packages/upgrade/test/common/downgrade_component_adapter_spec.ts [hunk 4]
```diff
@@ -10,9 +10,9 @@ import {TestBed, getTestBed, inject} from '@angular/core/testing';
 import * as angular from '@angular/upgrade/src/common/angular1';
 import {DowngradeComponentAdapter, groupNodesBySelector} from '@angular/upgrade/src/common/downgrade_component_adapter';
 
-import {nodes} from './test_helpers';
+import {nodes, testWithDifferentNg1Versions} from './test_helpers';
 
-{
+testWithDifferentNg1Versions(() => {
```
Comment (petebacondarwin) [inline packages/upgrade/test/common/downgrade_component_adapter_spec.ts]: I would rename this function to: `withEachNg1Version(...)`

File: packages/upgrade/test/common/test_helpers.ts [hunk 5]
```diff
@@ -5,6 +5,57 @@
...
+
+        // Remove AngularJS to leave a clean state for subsequent tests.
+        setNg1(undefined);
+        delete (window as any).angular;
+      });
+
+      specSuite();
+    });
+  });
+}
```
Comment (petebacondarwin) [inline packages/upgrade/test/common/test_helpers.ts]: Nice helper function!

**Review by petebacondarwin — APPROVED** [review]
LGTM Although I would consider the helper name change

File: packages/upgrade/src/common/angular1.ts (hunk 2 above)
Comment (gkalpak) [reply packages/upgrade/src/common/angular1.ts]: No 😕 (But it allows us to manually update it in `setAngularJSGlobal()`.)

Comment (petebacondarwin) [reply packages/upgrade/src/common/angular1.ts]: Oh yeah. I missed that.

File: packages/upgrade/test/BUILD.bazel (hunk 3 above)
Comment (gkalpak) [reply packages/upgrade/test/BUILD.bazel]: I don't think so. I've copied this from other places (following @alexeagle's advice) and I don't think we have issues for all of them. It is an implicit TODO as part of making all packages "buildable" with bazel, I guess :grin:

Comment (alexeagle) [reply packages/upgrade/test/BUILD.bazel]: Yes this is consistent with other tests.

**Review by alexeagle — APPROVED** [review]
Approval for bazel (didn't read all the changes)

**Comment by vicb** [comment]
[presubmit](http://test/OCL:186820654:BASE:186820680:1519423780216:b42d20c8)

File: packages/platform-browser/test/security/html_sanitizer_spec.ts [hunk 6]
```diff
@@ -149,7 +149,6 @@ import {sanitizeHtml} from '../../src/security/html_sanitizer';
     // See https://github.com/cure53/DOMPurify/releases/tag/0.6.7
     it('should not allow JavaScript hidden in badly formed HTML to get through sanitization (Firefox bug)',
        () => {
-         debugger;
```
Comment (IgorMinar) [inline packages/platform-browser/test/security/html_sanitizer_spec.ts]: ouch

File: package.json [hunk 7]
```diff
@@ -47,9 +47,10 @@
     "@types/selenium-webdriver": "3.0.7",
     "@types/source-map": "^0.5.1",
     "@types/systemjs": "0.19.32",
-    "angular": "1.5.0",
-    "angular-animate": "1.5.0",
-    "angular-mocks": "1.5.0",
+    "angular": "npm:angular@1.6",
+    "angular-1.5": "npm:angular@1.5",
```
Comment (IgorMinar) [inline package.json]: oh wow. I had no idea this was possible. hats off to you!

**Review by IgorMinar — APPROVED** [review]

File: package.json (hunk 7 above)
Comment (IgorMinar) [reply package.json]: that makes me think that we could actually use this trick to rename angular to angularjs in node_modules. it's just a nit. I don't insist on you changing it now, but it's an interesting option that I didn't realize we had until now.

Comment (petebacondarwin) [reply package.json]: we've been using this to test against multiple jquery versions in AngularJS for a while now. (This allowed us to rid ourselves of a whole load of bower deps).

Comment (gkalpak) [reply package.json]: Yeah, I learned this from @mgol :smiley: https://github.com/angular/angular.js/commit/ff2ce9dddec73601db727361812429180798205b

**Pull request closed without merging**